lib/os/onoff.c

```diff
@@ -600,3 +600,51 @@ int onoff_monitor_unregister(struct onoff_manager *mgr,
 
 	return rv;
 }
+
+int onoff_sync_lock(struct onoff_sync_service *srv,
+		    k_spinlock_key_t *keyp)
+{
+	*keyp = k_spin_lock(&srv->lock);
+	return srv->count;
+}
+
+int onoff_sync_finalize(struct onoff_sync_service *srv,
+			k_spinlock_key_t key,
+			struct onoff_client *cli,
+			int res,
+			bool on)
+{
+	uint32_t state;
+
+	/* Clear errors visible when locked.  If they are to be
+	 * preserved the caller must finalize with the previous
+	 * error code.
+	 */
+	if (srv->count < 0) {
+		srv->count = 0;
+	}
+	if (res < 0) {
+		srv->count = res;
+		state = ONOFF_STATE_ERROR;
+	} else if (on) {
+		srv->count += 1;
+		state = ONOFF_STATE_ON;
+	} else {
+		srv->count -= 1;
+		state = ONOFF_STATE_OFF;
```

Comment:
should it go to `OFF` state only when count reaches 0?

Reply:
Yes.  However if control flow reaches this point `state` isn't used in what follows.  I'll replace this with a comment.